src/main/python/mmlspark/cyber/ml/access_anomalies/collaborative_filtering.py

```diff
@@ -0,0 +1,721 @@
+
+        if not self.algo_cf_params.implicit:
+            complementset_factor = self.algo_cf_params.complementset_factor
+            neg_score = self.algo_cf_params.neg_score
+            assert complementset_factor is not None and neg_score is not None
+
+            comp_df = ComplementAccessTransformer(
+                tenant_col, [indexed_user_col, indexed_res_col], complementset_factor
+            ).transform(indexed_df).withColumn(
+                scaled_likelihood_col,
+                f.lit(neg_score)
+            )
+        else:
+            comp_df = None
+
+        scaled_df = self._get_scaled_df(indexed_df).select(
+            tenant_col, indexed_user_col, indexed_res_col, scaled_likelihood_col
+        )
+
+        return scaled_df.union(comp_df) if comp_df is not None else scaled_df
+
+    @timefunc
+    def _train_cf(self, als: ALS, df: DataFrame) -> Tuple[DataFrame, DataFrame]:
+        tenant_col = self.tenant_col
+        indexed_user_col = self.indexed_user_col
+        user_vec_col = self.user_vec_col
+        indexed_res_col = self.indexed_res_col
+        res_vec_col = self.res_vec_col
+
+        spark_model = als.fit(df)
+
+        user_model_df = spark_model.userFactors.select(
+            f.col('id').alias(indexed_user_col),
+            f.col('features').alias(user_vec_col)
+        ).join(
+            df.select(indexed_user_col, tenant_col).distinct(), indexed_user_col
+        ).select(
+            tenant_col, indexed_user_col, user_vec_col
+        )
+
+        res_model_df = spark_model.itemFactors.select(
+            f.col('id').alias(indexed_res_col),
+            f.col('features').alias(res_vec_col)
+        ).join(
+            df.select(indexed_res_col, tenant_col).distinct(), indexed_res_col
+        ).select(
+            tenant_col, indexed_res_col, res_vec_col
+        )
+
+        return user_model_df, res_model_df
+
+    @timefunc
+    def create_spark_model_vectors_df(self, df: DataFrame) -> UserResourceCfDataframeModel:
+        tenant_col = self.tenant_col
+        indexed_user_col = self.indexed_user_col
+        user_vec_col = self.user_vec_col
+        indexed_res_col = self.indexed_res_col
+        res_vec_col = self.res_vec_col
+        max_iter = self.max_iter
+        distinct_tenants = df.select(tenant_col).distinct().cache()
+        num_tenants = distinct_tenants.count()
+        separate_tenants = self.separate_tenants
+        num_blocks = self.num_blocks if self.num_blocks is not None else (num_tenants if not separate_tenants else 10)
+
+        als = ALS(
+            rank=self.rank_param,
+            maxIter=max_iter,
+            regParam=self.reg_param,
+            numUserBlocks=num_blocks,
+            numItemBlocks=num_blocks,
+            implicitPrefs=self.algo_cf_params.implicit,
+            userCol=self.indexed_user_col,
+            itemCol=self.indexed_res_col,
+            ratingCol=self.scaled_likelihood_col,
+            nonnegative=True,
```

Comment:
do these two parameters need to be off-limits to users?

Reply:
Can you clarify the question here?